[export] Fix sym_bool serialization

yiming0416 opened the pull request:
Summary: 
When there is a `torch._check()` that checks if a sym_int is equal to some constant, it will generate 3 nodes in the graph with target `operation.ge`, `operator.le` and `operator.eq`. These operators belong to `_SYM_BOOL_OPS` but the `meta_val` of these nodes are are `bool` instead of `torch.SymBool`.

Similar things can happen to `torch.SymInt`, where a `node.target` belongs to `_SYM_INT_OPS` but `node.meta["val"]` is an `int` instead of `torch.SymInt`.

Therefore, we need to check both `meta_val` type and `node.target` type during serialization.

Test Plan: 
```
buck2 run @mode/dev-nosan caffe2/test:test_export -- -r test_sym_bool_torch_check_equal
buck2 run @mode/dev-nosan caffe2/test:test_export -- -r test_sym_int_torch_check_equal
```

Differential Revision: D67883754

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -551,11 +551,11 @@ def handle_call_function(self, node: torch.fx.Node):
-            if isinstance(meta_val, torch.SymInt):
-                sym_output = Argument.create(as_sym_int=self.serialize_sym_int_output(node.name, meta_val))
-            elif isinstance(meta_val, torch.SymFloat):
+            if isinstance(meta_val, torch.SymFloat) or node.target in _SYM_FLOAT_OPS:
```
The order of these checks is not significant, right? Are `_SYM_X_OPS` for X in `FLOAT`, `INT`, `BOOL` disjoint?

Can you comment why you needed to check the ops? Is `meta_val` sometimes None or does it not match the `torch.SymX` types?

yiming0416 replied:
@avikchaudhuri
The order of these checks matter. Although `_SYM_X_OPS` are disjoint, consider the line 
```
x / torch.sym_sqrt(x.shape[0])
```
in test `test_sym_sqrt`, the op is `torch.sym_sqrt` belongs to `_SYM_INT_OPS`, but the `meta_val` is `torch.SymFloat`. That's why float has to go first.

The reason I came up with this fix is that #136364 caused regression on serializing internal models reflected by the internal dashboard. Because for some internal models, there could be a node with target `operator.ge` but has a meta_val of type `bool` instead of `torch.SymBool`. I tried to add tests but I was unable to reproduce this behavior in OSS. Any thoughts on this? 

Also before #136364, we check both meta_val and op.

yiming0416 replied:
@avikchaudhuri Please see the updated summary and added test.

avikchaudhuri replied:
I'm puzzled why `torch.sym_sqrt` belongs to `_SYM_INT_OPS` (Is it because the expected input type is symint but here we're talking about output type? it looks like a floaty op to me.) @angelayi 

OK otherwise.

avikchaudhuri reviewed (approved):
Can you add some tests for each case of float, int, bool to exercise the new checks of ops?

yiming0416 committed:
[export] Fix sym_bool serialization (#144295)

Summary:

When there is a `torch._check()` that checks if a sym_int is equal to some constant, it will generate 3 nodes in the graph with target `operation.ge`, `operator.le` and `operator.eq`. These operators belong to `_SYM_BOOL_OPS` but the `meta_val` of these nodes are are `bool` instead of `torch.SymBool`.

Similar things can happen for `torch.SymInt` as well, where a `node.target` belongs to `_SYM_INT_OPS` but `node.meta["val"]` is an `int` instead of `torch.SymInt`.

Therefore, we need to check both `meta_val` type and `node.target` type during serialization.

Test Plan:
```
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_sym_bool_torch_check_equal
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_sym_int_torch_check_equal
```

Reviewed By: avikchaudhuri, angelayi

Differential Revision: D67883754
